cpp/src/arrow/compute/kernels/random.cc

```diff
@@ -0,0 +1,85 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <random>
+
+#include "arrow/builder.h"
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/pcg_random.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+namespace {
+
+// Generates a random floating point number in range [0, 1).
+double generate_uniform(random::pcg64& rng) {
+  return (rng() >> 11) * (1.0 / 9007199254740992.0);
+}
```

Comment:
We can add some checks and explanations here in case the code rotten.
```
// This equation is copied from numpy. It calculates `rng() / 2^64` and
// the return value is strictly less than 1.
static_assert(random::pcg64_fast::min() == 0ULL, "");
static_assert(random::pcg64_fast::max() == ~0ULL, "");
return (rng() >> 11) * (1.0 / 9007199254740992.0);
......
```